common/unified/matrix/csr_kernels.cpp

```diff
@@ -54,53 +54,71 @@ namespace GKO_DEVICE_NAMESPACE {
+                     const IndexType* perm,
+                     const matrix::Csr<ValueType, IndexType>* orig,
+                     matrix::Csr<ValueType, IndexType>* col_permuted)
 {
+    auto num_rows = orig->get_size()[0];
+    auto nnz = orig->get_num_stored_elements();
+    auto size = std::max(num_rows, nnz);
     run_kernel(
         exec,
-        [] GKO_KERNEL(auto tid, auto permutation, auto inv_permutation) {
-            inv_permutation[permutation[tid]] = tid;
+        [] GKO_KERNEL(auto tid, auto num_rows, auto num_nonzeros,
+                      auto permutation, auto in_row_ptrs, auto in_col_idxs,
+                      auto in_vals, auto out_row_ptrs, auto out_col_idxs,
+                      auto out_vals) {
+            if (tid < num_nonzeros) {
+                out_col_idxs[tid] = permutation[in_col_idxs[tid]];
+                out_vals[tid] = in_vals[tid];
+            }
+            if (tid <= num_rows) {
+                out_row_ptrs[tid] = in_row_ptrs[tid];
+            }
         },
-        size, permutation_indices, inv_permutation);
+        size, num_rows, nnz, perm, orig->get_const_row_ptrs(),
+        orig->get_const_col_idxs(), orig->get_const_values(),
+        col_permuted->get_row_ptrs(), col_permuted->get_col_idxs(),
+        col_permuted->get_values());
 }
 
-GKO_INSTANTIATE_FOR_EACH_INDEX_TYPE(GKO_DECLARE_INVERT_PERMUTATION_KERNEL);
+GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
+    GKO_DECLARE_CSR_INV_COL_PERMUTE_KERNEL);
 
 
 template <typename ValueType, typename IndexType>
-void inverse_column_permute(std::shared_ptr<const DefaultExecutor> exec,
-                            const IndexType* perm,
-                            const matrix::Csr<ValueType, IndexType>* orig,
-                            matrix::Csr<ValueType, IndexType>* column_permuted)
+void inv_col_scale_permute(std::shared_ptr<const DefaultExecutor> exec,
+                           const ValueType* scale, const IndexType* perm,
+                           const matrix::Csr<ValueType, IndexType>* orig,
+                           matrix::Csr<ValueType, IndexType>* col_permuted)
 {
     auto num_rows = orig->get_size()[0];
     auto nnz = orig->get_num_stored_elements();
     auto size = std::max(num_rows, nnz);
     run_kernel(
         exec,
-        [] GKO_KERNEL(auto tid, auto num_rows, auto num_nonzeros,
+        [] GKO_KERNEL(auto tid, auto num_rows, auto num_nonzeros, auto scale,
                       auto permutation, auto in_row_ptrs, auto in_col_idxs,
                       auto in_vals, auto out_row_ptrs, auto out_col_idxs,
                       auto out_vals) {
             if (tid < num_nonzeros) {
-                out_col_idxs[tid] = permutation[in_col_idxs[tid]];
-                out_vals[tid] = in_vals[tid];
+                const auto out_col = permutation[in_col_idxs[tid]];
+                out_col_idxs[tid] = out_col;
+                out_vals[tid] = in_vals[tid] / scale[out_col];
```

Comment:
the second formula $P^{-1} (P S^{-1} P^T)$ is only useful for inverting a scaled permutation into the $PS$ form again, we don't need it for the scaling. Is there a question on this kernel?

Reply:
It is only to clear my understanding is correct or not.
in this kernel, you did the permutation first and then the scaling directly not the other direction because you do not have and do not need to generate the second formula